pkg/controller/direct/datacatalog/tag_controller.go

```diff
@@ -0,0 +1,380 @@
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Resolve the tag template reference first to get the full template name.
+	tagTemplateName, err := obj.Spec.TemplateRef.NormalizedExternal(ctx, reader, obj.GetNamespace())
+	if err != nil {
+		return nil, fmt.Errorf("resolving template reference: %w", err)
+	}
+
+	// DataCatalogTagIdentity requires the parent KRM object (Entry or EntryGroup)
+	// and the TagTemplate name to construct the potential tag name.
+	// However, the actual Tag ID is server-generated. We'll use the identity mostly
+	// for the parent resource name.
+	id, err := krm.NewTagIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	projectID := id.Parent().ProjectID
+
+	gcpClient, err := m.client(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tagAdapter{
+		gcpClient:       gcpClient,
+		id:              id, // Primarily used for parent resolution
+		desired:         obj,
+		reader:          reader,
+		tagTemplateName: tagTemplateName, // Store resolved template name
+	}, nil
+}
+
+func (m *tagModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type tagAdapter struct {
+	gcpClient       *api.Client
+	id              *krm.TagIdentity // Primarily for parent resolution
+	desired         *krm.DataCatalogTag
+	actual          *pb.Tag
+	reader          client.Reader
+	tagTemplateName string // Resolved TagTemplate resource name
+}
+
+var _ directbase.Adapter = &tagAdapter{}
+
+// Find implements directbase.Adapter.
+// Data Catalog Tags don't have a Get RPC. We list tags on the parent and filter.
+func (a *tagAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	mapCtx := &direct.MapContext{}
+
+	parentName := a.id.Parent().String()
+	log.V(2).Info("listing datacatalog tags", "parent", parentName)
+
+	// Get desired scope for matching
+	desiredProto := DataCatalogTagSpec_ToProto(mapCtx, &a.desired.Spec)
+	if mapCtx.Err() != nil {
+		return false, mapCtx.Err()
+	}
+	desiredColumn := desiredProto.GetColumn() // Works even if nil
+
+	// Use the fully resolved template name for matching
+	desiredTemplate := a.tagTemplateName
+
+	req := &pb.ListTagsRequest{Parent: parentName}
+	it := a.gcpClient.ListTags(ctx, req)
+	for {
+		tag, err := it.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			// PermissionDenied might occur if parent doesn't exist
+			if direct.IsPermissionDenied(err) {
+				log.V(1).Info("permission denied listing tags, parent likely missing", "parent", parentName, "error", err)
+				return false, nil // Treat as not found
+			}
+			return false, fmt.Errorf("listing datacatalog tags for parent %q: %w", parentName, err)
+		}
+
+		// Match based on template and scope (column)
+		if tag.Template == desiredTemplate {
```

Comment:
I am a bit surprised we don't use "resourceID" to acquire a GCP resource?

Reply:
it should be resourceID. Fixed it.